### PR TITLE
Sparse tensor printing; add NotImplemented autograd fn

### LINKED ISSUE #9412: grad_fn of sparse tensor not displayed correctly
The grad_fn are shown as Error when print a sparse tensor that requires_grad
```
a = torch.sparse.FloatTensor(
    torch.LongTensor([[0, 0, 1],
                      [0, 1, 1]]),
    torch.FloatTensor([2, 3, 4]),
    torch.Size([2, 2])).requires_grad_()

b = torch.sparse.FloatTensor(
    torch.LongTensor([[0, 0, 1],
                      [0, 1, 1]]),
    torch.FloatTensor([1, 1, 1]),
    torch.Size([2, 2])).requires_grad_()

d = a.mul(b)
d.backward(a)

print(d.grad_fn) # <SNativeMulBackward object at 0x7efc103176d0>
print(d._indices().grad_fn) # <Error object at 0x7efc10317a50>
print(d._values().grad_fn)  # <Error object at 0x7efc103176d0> 



print(d)
----------
torch.sparse.FloatTensor of size (2,2) with indices:
tensor([[0, 0, 1],
        [0, 1, 1]], grad_fn=<Error>)
and values:
tensor([2., 3., 4.], grad_fn=<Error>)



print(d.to_dense())
-------------------
tensor([[2., 3.],
        [0., 4.]], grad_fn=<Error>)
```

Related codes are [here](https://github.com/pytorch/pytorch/blob/2ead3b0e54c0e53d3ab20c0b87a4aea873f6172d/torch/_tensor_str.py#L223-L227). How should I fix this? cc @colesbury 

EDIT: `print(d.to_dense())` is legit since there is not grad_fn defined for `to_dense`, `print(d._indices().grad_fn)` and `print(d._values().grad_fn)` are probably legit too since their grad_fn are not defined.

### PR DESCRIPTION
**Pull request by ssnl**
Commits:

1. Add autograd function `NotImplemented` (subclass of `Error`) so python `grad_fn` prints nicer. Since `Error` is used in `DelayedError` to implement `@oncedifferentiable`, I can't just change its name. cc @colesbury 

2. Add printing for sparse tensors. Fixes https://github.com/pytorch/pytorch/issues/9412 . cc @weiyangfb @li-roy . 

3. Add tests for sparse printing

Examples:
```diff
  In [2]: x = torch.sparse.FloatTensor(torch.arange(4).view(2,2), torch.randn(2, 2), [10, 10, 2])
  
  In [3]: x
  Out[3]:
- torch.sparse.FloatTensor of size (10,10,2) with indices:
- tensor([[0, 1],
-         [2, 3]])
- and values:
- tensor([[-1.1832, -0.5927],
-         [ 0.0831,  0.2511]])
+ tensor(indices=tensor([[0, 1],
+                        [2, 3]]),
+        values=tensor([[ 1.5081,  0.3451],
+                       [-0.0392,  0.4776]]),
+        size=(10, 10, 2), nnz=2, layout=torch.sparse_coo)
  
  In [4]: x.requires_grad_()
  Out[4]:
- torch.sparse.FloatTensor of size (10,10,2) with indices:
- tensor([[0, 1],
-         [2, 3]], grad_fn=<Error>)
- and values:
- tensor([[-1.1832, -0.5927],
-         [ 0.0831,  0.2511]], grad_fn=<Error>)
+ tensor(indices=tensor([[0, 1],
+                        [2, 3]]),
+        values=tensor([[ 1.5081,  0.3451],
+                       [-0.0392,  0.4776]]),
+        size=(10, 10, 2), nnz=2, layout=torch.sparse_coo, requires_grad=True)
  
  In [5]: x + x
  Out[5]:
- torch.sparse.FloatTensor of size (10,10,2) with indices:
- tensor([[0, 1],
-         [2, 3]], grad_fn=<Error>)
- and values:
- tensor([[-2.3664, -1.1855],
-         [ 0.1662,  0.5021]], grad_fn=<Error>)
+ tensor(indices=tensor([[0, 1],
+                        [2, 3]]),
+        values=tensor([[ 3.0162,  0.6902],
+                       [-0.0785,  0.9553]]),
+        size=(10, 10, 2), nnz=2, layout=torch.sparse_coo, grad_fn=<AddBackward0>)
  
  In [6]: x.double()
  Out[6]:
- torch.sparse.DoubleTensor of size (10,10,2) with indices:
- tensor([[0, 1],
-         [2, 3]], grad_fn=<Error>)
- and values:
- tensor([[-1.1832, -0.5927],
-         [ 0.0831,  0.2511]], dtype=torch.float64, grad_fn=<Error>)
+ tensor(indices=tensor([[0, 1],
+                        [2, 3]]),
+        values=tensor([[ 1.5081,  0.3451],
+                       [-0.0392,  0.4776]]),
+        size=(10, 10, 2), nnz=2, dtype=torch.float64, layout=torch.sparse_coo,
+        grad_fn=<NotImplemented>)
  
  In [7]: x = torch.sparse.FloatTensor(torch.ones(0, 2, dtype=torch.long), torch.randn(2, 0), [0])
  
  In [8]: x
  Out[8]:
- torch.sparse.FloatTensor of size (0,) with indices:
- tensor([], size=(0, 2), dtype=torch.int64)
- and values:
- tensor([], size=(2, 0))
+ tensor(indices=tensor([], size=(0, 2)),
+        values=tensor([], size=(2, 0)),
+        size=(0,), nnz=2, layout=torch.sparse_coo)
  
  In [9]: x = torch.sparse.FloatTensor(torch.ones(0, 2, dtype=torch.long), torch.randn(2), [])
  
  In [10]: x
  Out[10]:
- torch.sparse.FloatTensor of size () with indices:
- tensor([], size=(0, 2), dtype=torch.int64)
- and values:
- tensor([-0.0064,  0.8518])
+ tensor(indices=tensor([], size=(0, 2)),
+        values=tensor([ 0.9800, -0.5978]),
+        size=(), nnz=2, layout=torch.sparse_coo)
```

### CODE REVIEW

File: torch/_tensor_str.py [hunk 1]
```diff
@@ -221,23 +221,50 @@ def get_summarized_data(self):
...
     summarize = self.numel() > PRINT_OPTS.threshold
 
     suffix = ''
-    if not torch._C._is_default_type_cuda():
-        if self.device.type == 'cuda':
-            suffix += ', device=\'' + str(self.device) + '\''
-    else:
-        if self.device.type == 'cpu' or torch.cuda.current_device() != self.device.index:
-            suffix += ', device=\'' + str(self.device) + '\''
+    if not skip_device:
```
Comment (weiyangfb) [inline torch/_tensor_str.py]: Does the device info get printed for sparse? Maybe add a test for it?

File: test/test_sparse.py [hunk 2]
```diff
@@ -107,6 +107,37 @@ def randn(self, *args, **kwargs):
...
+            raise unittest.SkipTest("test_print is the same regardless of is_uncoalesced")
+        for nnz in (0, 3):
+            indices = torch.arange(2 * nnz).view(2, nnz).type(self.IndexTensor)
+            if indices.numel() > 0:
+                indices[:, -1] = indices[:, 0]  # make it uncoalesced
+            values = torch.arange(3 * nnz).view(nnz, 3).type(self.ValueTensor)
+            values.add_(torch.randn_like(values) * 0.1)
+            sp_tensor = self.SparseTensor(indices, values, (200, 200, 3))
+
+            def get():
```
Comment (weiyangfb) [inline test/test_sparse.py]: maybe put the closure outside of for-loop and under test_print()?

**Review by weiyangfb — COMMENTED** [review]
Very nice!

**Review by gchanan — COMMENTED** [review]
Could you include before and after comparisons?

Without checking the before-and-after closely, this looks good to me (after the nits are addressed).  I have some concerns about how this looks post https://github.com/pytorch/pytorch/pull/9279 -- for example, you never print the shape of the indices tensor, but your print can be wrong in the numel() == 0 case, because the shape is really part of the tensor print (because you can't deduce the shape from the print).

**Comment by fmassa** [comment]
ping @SsnL on Greg's comments. Also, this now needs a rebase

File: aten/src/ATen/native/sparse/SparseUtils.h [hunk 3]
```diff
@@ -108,16 +108,9 @@ inline LongTensor _newFlattenedIndices(const SparseTensor& self, bool forceClone
 // TODO: Expose this for real in ATen, some day?
 // NB: Doesn't preserve data.
 inline Tensor _new_values_with_size_of(const Tensor& values, int64_t nnz) {
-  if (values.numel() == 0) { // values tensor uninitialized
```
Comment (SsnL) [inline aten/src/ATen/native/sparse/SparseUtils.h]: I believe this should be removed after @yf225 's patch.

Comment (yf225) [reply aten/src/ATen/native/sparse/SparseUtils.h]: Yes I have a PR coming up that will also make the same change in this function. Feel free to fix it in this PR first :)

**Review by gchanan — APPROVED** [review]

File: torch/_tensor_str.py [hunk 4]
```diff
@@ -222,50 +240,60 @@ def get_summarized_data(self):
...
-        tensor_str = '[]'
+    has_default_dtype = self.dtype == torch.get_default_dtype() or self.dtype == torch.int64
+
+    if self.is_sparse:
+        suffixes.append('size=' + str(tuple(self.shape)))
+        suffixes.append('nnz=' + str(self._nnz()))
+        if not has_default_dtype:
+            suffixes.append('dtype=' + str(self.dtype))
+        indices_prefix = 'indices=tensor('
+        indices = self._indices().detach()
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/_tensor_str.py]: Apparently we don't check returned dtype for setting that flag. See
```py
>>> x = torch.sparse.FloatTensor(torch.ones(0, 2, dtype=torch.long), torch.randn(2, 0
, [0])
>>> x.requires_grad_()
tensor(indices=tensor([], size=(0, 2)),
       values=tensor([], size=(2, 0)),
       size=(0,), nnz=2, layout=torch.sparse_coo, requires_grad=True)
>>> x._indices()
tensor([], size=(0, 2), dtype=torch.int64, grad_fn=<NotImplemented>)
```

I'll fixing this in a new PR.

**Pull request closed without merging**